data-plane/core/session/src/session_receiver.rs

```diff
@@ -0,0 +1,1772 @@
+// Copyright AGNTCY Contributors (https://github.com/agntcy)
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
+
+use slim_datapath::messages::utils::SlimHeaderFlags;
+use slim_datapath::{
+    api::{ProtoMessage as Message, SessionHeader, SlimHeader},
+    messages::Name,
+};
+use tokio::sync::mpsc::{Receiver, Sender};
+use tracing::{debug, error};
+
+use crate::{
+    MessageDirection, SessionError, SessionType, Transmitter,
+    common::SessionMessage,
+    receiver_buffer::ReceiverBuffer,
+    timer::Timer,
+    timer_factory::{TimerFactory, TimerSettings},
+};
+
+const DRAIN_TIMER_ID: u32 = 0;
+
+// structs used in the pending rtx map
+struct PendingRtxVal {
+    timer: Timer,
+    message: Message,
+}
+
+struct PendingRtxKey {
+    name: Name,
+    id: u32,
+}
+
+impl PartialEq for PendingRtxKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name && self.id == other.id
+    }
+}
+
+impl Eq for PendingRtxKey {}
+
+impl Hash for PendingRtxKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.id.hash(state);
+    }
+}
+
+#[allow(dead_code)]
+struct SessionReceiverInternal<T>
+where
+    T: Transmitter + Send + Sync + Clone + 'static,
+{
+    /// buffer with received packets one per endpoint
+    buffer: HashMap<Name, ReceiverBuffer>,
+
+    /// list of pending RTX requests per name/id
+    pending_rtxs: HashMap<PendingRtxKey, PendingRtxVal>,
+
+    /// timer factory to crate timers for rtx
+    /// if None, no rtx is sent. In this case there is no
+    /// ordered delivery to the app and messages are sent
+    /// as soon as they arrive at there receiver without using
+    /// the buffer
+    timer_factory: Option<TimerFactory>,
+
+    /// session id where to send the messages
+    session_id: u32,
+
+    /// local name to use as source for the rtx messages
+    local_name: Name,
+
+    /// session type
+    session_type: SessionType,
+
+    /// if true send acks for each received message
+    send_acks: bool,
+
+    /// send to slim/app
+    tx: T,
+
+    /// received for signals
+    rx: Receiver<SessionMessage>,
+
+    /// drain state - when true, no new messages from app are accepted
+    is_draining: bool,
+
+    /// drain timer - when set, we're waiting for pending acks during grace period
+    drain_timer: Option<Timer>,
+}
+
+#[allow(dead_code)]
+impl<T> SessionReceiverInternal<T>
+where
+    T: Transmitter + Send + Sync + Clone + 'static,
+{
+    /// to create the timer factory and send rtx messages
+    /// timer_settings and tx_timer must be not null
+    #[allow(clippy::too_many_arguments)]
+    fn new(
+        timer_settings: Option<TimerSettings>,
+        session_id: u32,
+        local_name: Name,
+        session_type: SessionType,
+        send_acks: bool,
+        tx: T,
```

Comment:
Do we need this transmitter to be not optional? For instance in the control plane we won't have a transmitter 🤔